sled-agent/config-reconciler/src/dataset_serialization_task.rs

```diff
@@ -399,9 +456,140 @@ impl DatasetTask {
+                    // dataset. (This includes non-crucible Omicron zone durable
+                    // datasets.)
+                    format!("{zpool}/{CRYPT_DATASET}"),
+                    // We _don't_ need to check children of
+                    // `{zpool}/{ZONE_DATASET}`: these are all transient, and
+                    // should be destroyed/created on demand.
+                ]
+            })
+            .collect::<Vec<_>>();
+
+        for properties in zfs
+            .get_dataset_properties(
+                &datasets_of_interest,
+                WhichDatasets::SelfAndChildren,
+            )
+            .await?
+        {
+            // Skip any errors parsing dataset names: we expect to get some
+            // errors! We won't be able to parse any dataset names that can't be
+            // represented by `DatasetKind`, which will always include at least
+            // the two parent datasets we asked for above (`zpool` and
+            // `zpool/crypt`).
+            //
+            // Our goal here is to prune datasets that _we_ created; we only
+            // create datasets that have representable `DatasetName`s, so this
+            // also acts as a safeguard that we should never delete datasets we
+            // don't fully understand.
+            let Ok(dataset) = DatasetName::from_str(&properties.name) else {
+                continue;
+            };
+
+            // Does this dataset have an ID set? If we created it, we expect it
+            // does, and we can easily check whether it still exists in our
+            // config. A dataset with a known `DatasetKind` without an ID set is
+            // pretty unexpected: we'll search through our config datasets to
+            // find a matching name; if we do, we'll assume it's that one.
+            let present_in_config = match properties.id {
+                Some(id) => datasets.contains_key(&id),
+                None => {
+                    if datasets.iter().any(|d| d.name == dataset) {
+                        warn!(
+                            self.log,
+                            "found on-disk dataset without an ID \
+                             that matches a config dataset by name";
+                            "dataset" => dataset.full_name(),
+                        );
+                        true
+                    } else {
+                        warn!(
+                            self.log,
+                            "found on-disk dataset without an ID \
+                             that doesn't match any config datasets; assuming \
+                             it should be marked as an orphan";
+                            "dataset" => dataset.full_name(),
+                        );
+                        false
+                    }
+                }
+            };
+            if present_in_config {
+                continue;
+            }
```

Comment:
Up to you, but, this whole block *could* be marked as "for backwards compatibility" -- we could someday remove it when we eventually are confident all datasets have UUIDs

Reply:
I'm not sure we can guarantee that - isn't "create the dataset" a separate step from "set the `oxide:uuid` property on a dataset" (and therefore we could succeed in creating but fail in setting the ID)?

And maybe a step beyond - even if that's wrong and we can create+set ID atomically, I don't think we have a reasonable way to represent reading ZFS datasets where an ID is required (i.e., I'd expect `zfs.get_dataset_properties()` to _always_ return an `Option<DatasetUuid>`, which we have to handle somehow).